### PR TITLE
MLIR: Updating CMakeLists and adding FindMLIR.cmake

### PR DESCRIPTION
**Pull request by Robertorosmaninho**
This updates will allow LDC to build the code that will allow the support for MLIR

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -348,6 +348,94 @@ endif()
...
+${LLVM_LIBRARY_DIRS}/libMLIRTestDialect.${suffix}
+${LLVM_LIBRARY_DIRS}/libMLIRTestIR.${suffix}
+${LLVM_LIBRARY_DIRS}/libMLIRTestPass.${suffix}
+${LLVM_LIBRARY_DIRS}/libMLIRTestTransforms.${suffix}
+${LLVM_LIBRARY_DIRS}/libMLIRTransforms.${suffix}
+${LLVM_LIBRARY_DIRS}/libMLIRTransformUtils.${suffix}
+${LLVM_LIBRARY_DIRS}/libMLIRTranslateClParser.${suffix}
+${LLVM_LIBRARY_DIRS}/libMLIRTranslation.${suffix}
+${LLVM_LIBRARY_DIRS}/libMLIRVectorOps.${suffix}
+${LLVM_LIBRARY_DIRS}/libMLIRVectorToLLVM.${suffix}
```
Comment (thewilsonator) [inline CMakeLists.txt]: Is it possible to do this with a glob?

File: CMakeLists.txt [hunk 2]
```diff
@@ -348,6 +348,94 @@ endif()
 message(STATUS "LDC version identifier: ${LDC_VERSION}")
 configure_file(driver/ldc-version.cpp.in driver/ldc-version.cpp)
 
+#
+# Setting MLIR headers
+#
+
+find_package(MLIR REQUIRED)
```
Comment (thewilsonator) [inline CMakeLists.txt]: please remove the `REQUIRED`

**Comment by kinke** [comment]
I really hope this can be substantially simplified; a list of 50 (!) libs to be added is just horrible (I don't think we need all of them anyway). I probably won't bother with LLVM 10 until it's really finished, which might take another month.

File: CMakeLists.txt (hunk 1 above)
Comment (thewilsonator) [reply CMakeLists.txt]: Just use `MLIR_LIBRARIES` in please of `MLIR_LIB`.

**Comment by thewilsonator** [comment]
Fixed that. This should be good to squash and merge.

File: CMakeLists.txt (hunk 2 above)
Comment (Robertorosmaninho) [reply CMakeLists.txt]: thanks

Comment (Robertorosmaninho) [reply CMakeLists.txt]: thanks

**Comment by kinke** [comment]
This could have waited until the imminent release is over. - I'll create a new branch for it then.

File: cmake/Modules/FindMLIR.cmake [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+set(MLIR_LIB_DIRS ${MLIR_LIB_DIR}/..)
+
+if(EXISTS ${MLIR_ROOT_DIR})
+#Resources to automatically generate Ops.h.inc and Ops.cpp.inc, assuming that
+#mlir-tblgen is already built and is on the usual directory at llvm dir
+set(MLIR_MAIN_SRC_DIR ${MLIR_INCLUDE_DIR}/.. )
+set(MLIR_TABLEGEN_EXE ${LLVM_ROOT_DIR}/bin/mlir-tblgen)
+set(MLIR_INCLUDE_DIR ${MLIR_BUILD_INCLUDE_DIR}/..)
+
+function(mlir_tablegen)
```
Comment (kinke) [inline cmake/Modules/FindMLIR.cmake]: Is this function actually used somewhere?

Comment (Robertorosmaninho) [reply cmake/Modules/FindMLIR.cmake]: Yes, this function is necessary to automatically generate c++ files from a code that describes dialect operations written in tablegen. These dialects operations are the subject of my next PR of MLIR support.

**Comment by kinke** [comment]
This is apparently far from usable: with LLVM 10 RC2+, (included) MLIR isn't found on Linux and Mac (see the [Azure CI logs](https://dev.azure.com/ldc-developers/ldc/_build/results?buildId=1213&view=logs&j=d63d0aa7-5d01-5c9f-d5bd-b8a45119ae4e&t=e8d8503e-b7e2-5213-9d0c-cba4114111fa) from #3307). The output isn't very nice:
```
-- MLIR Dir: MLIR_ROOT_DIR-NOTFOUND
-- MLIR Include Dir: MLIR_INCLUDE_DIR-NOTFOUND/..
-- MLIR Lib Dir: MLIR_LIB_DIR-NOTFOUND/..
-- MLIR Build Include Dir: MLIR_BUILD_INCLUDE_DIR-NOTFOUND/..
```

On Windows, it's completely failing (same on my box):
```
-- MLIR Dir: C:/Program Files (x86)/Windows Kits/10/Include/10.0.18362.0/cppwinrt
-- MLIR Include Dir: C:/Program Files (x86)/Windows Kits/10/Include/10.0.18362.0/um/..
-- MLIR Lib Dir: MLIR_LIB_DIR-NOTFOUND/..
-- MLIR Build Include Dir: MLIR_BUILD_INCLUDE_DIR-NOTFOUND/..
-- Found MLIR: C:/Program Files (x86)/Windows Kits/10/Include/10.0.18362.0/cppwinrt  
-- Buiding with MLIR support
```

File: cmake/Modules/FindMLIR.cmake [hunk 4]
```diff
@@ -0,0 +1,68 @@
+# - Try to find MLIR project at LLVM
+#
+# The following are set after configuration is done:
+#   MLIR_FOUND
+#   MLIR_ROOT_DIRS
+#   MLIR_INCLUDE_DIRS
+#   MLIR_BUILD_INCLUDE_DIR
+project(ldc)
+
+find_path(MLIR_ROOT_DIR NAMES "LICENSE.TXT" HINTS ${LLVM_ROOT_DIR}/../mlir)
```
Comment (kinke) [inline cmake/Modules/FindMLIR.cmake]: This just looks for the first dir containing a LICENSE.TXT file, starting with an extremely unlikely hint (`LLVM_ROOT_DIR` is usually either `/usr` or `~/my-llvm`, and the installed MLIR includes and libs are in the usual LLVM dirs).

File: cmake/Modules/FindMLIR.cmake (hunk 3 above)
Comment (kinke) [reply cmake/Modules/FindMLIR.cmake]: It looks like these files might already be generated. These are the contents of my `include/mir/Dialect` subdirectory:
```
.:
total 52
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 AffineOps
-rw-r--r-- 1 Martin 197121 3254 Feb 15 01:34 CommonFolders.h
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 FxpMathOps
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 GPU
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 LLVMIR
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 Linalg
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 LoopOps
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 QuantOps
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 SDBM
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 SPIRV
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 StandardOps
-rw-r--r-- 1 Martin 197121 3603 Feb 15 01:34 Traits.h
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 Utils
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 VectorOps

./AffineOps:
total 88
-rw-r--r-- 1 Martin 197121 24448 Feb 15 02:37 AffineOps.cpp.inc
-rw-r--r-- 1 Martin 197121 27981 Feb 15 01:34 AffineOps.h
-rw-r--r-- 1 Martin 197121 14307 Feb 15 02:37 AffineOps.h.inc
-rw-r--r-- 1 Martin 197121 13590 Feb 15 01:34 AffineOps.td
-rw-r--r-- 1 Martin 197121  1101 Feb 15 01:34 AffineOpsBase.td

./FxpMathOps:
total 116
-rw-r--r-- 1 Martin 197121 73370 Feb 15 02:37 FxpMathOps.cpp.inc
-rw-r--r-- 1 Martin 197121   935 Feb 15 01:34 FxpMathOps.h
-rw-r--r-- 1 Martin 197121 20922 Feb 15 02:37 FxpMathOps.h.inc
-rw-r--r-- 1 Martin 197121 11723 Feb 15 01:34 FxpMathOps.td
-rw-r--r-- 1 Martin 197121  1309 Feb 15 01:34 Passes.h

./GPU:
total 108
-rw-r--r-- 1 Martin 197121  2981 Feb 15 01:34 GPUDialect.h
-rw-r--r-- 1 Martin 197121 46714 Feb 15 02:37 GPUOps.cpp.inc
-rw-r--r-- 1 Martin 197121 24215 Feb 15 02:37 GPUOps.h.inc
-rw-r--r-- 1 Martin 197121 24562 Feb 15 01:34 GPUOps.td
-rw-r--r-- 1 Martin 197121  1053 Feb 15 01:34 MemoryPromotion.h
-rw-r--r-- 1 Martin 197121   848 Feb 15 01:34 Passes.h

./LLVMIR:
total 592
-rw-r--r-- 1 Martin 197121  13417 Feb 15 02:38 LLVMConversions.inc
-rw-r--r-- 1 Martin 197121   7067 Feb 15 01:34 LLVMDialect.h
-rw-r--r-- 1 Martin 197121   2160 Feb 15 01:34 LLVMOpBase.td
-rw-r--r-- 1 Martin 197121 263206 Feb 15 02:38 LLVMOps.cpp.inc
-rw-r--r-- 1 Martin 197121 104756 Feb 15 02:38 LLVMOps.h.inc
-rw-r--r-- 1 Martin 197121  31287 Feb 15 01:34 LLVMOps.td
-rw-r--r-- 1 Martin 197121   6480 Feb 15 02:38 LLVMOpsEnums.cpp.inc
-rw-r--r-- 1 Martin 197121   4330 Feb 15 02:38 LLVMOpsEnums.h.inc
-rw-r--r-- 1 Martin 197121   3524 Feb 15 02:38 NVVMConversions.inc
-rw-r--r-- 1 Martin 197121   1142 Feb 15 01:34 NVVMDialect.h
-rw-r--r-- 1 Martin 197121  52104 Feb 15 02:38 NVVMOps.cpp.inc
-rw-r--r-- 1 Martin 197121  23434 Feb 15 02:38 NVVMOps.h.inc
-rw-r--r-- 1 Martin 197121   5698 Feb 15 01:34 NVVMOps.td
-rw-r--r-- 1 Martin 197121   1998 Feb 15 02:38 ROCDLConversions.inc
-rw-r--r-- 1 Martin 197121   1560 Feb 15 01:34 ROCDLDialect.h
-rw-r--r-- 1 Martin 197121  30829 Feb 15 02:38 ROCDLOps.cpp.inc
-rw-r--r-- 1 Martin 197121  15515 Feb 15 02:38 ROCDLOps.h.inc
-rw-r--r-- 1 Martin 197121   3902 Feb 15 01:34 ROCDLOps.td

./Linalg:
total 12
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 Analysis
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 EDSC
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 IR
-rw-r--r-- 1 Martin 197121 1688 Feb 15 01:34 Passes.h
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 Transforms
drwxr-xr-x 1 Martin 197121    0 Feb 15 15:29 Utils

./Linalg/Analysis:
total 8
-rw-r--r-- 1 Martin 197121 5801 Feb 15 01:34 DependenceAnalysis.h

./Linalg/EDSC:
total 16
-rw-r--r-- 1 Martin 197121 9245 Feb 15 01:34 Builders.h
-rw-r--r-- 1 Martin 197121  945 Feb 15 01:34 Intrinsics.h

./Linalg/IR:
total 200
-rw-r--r-- 1 Martin 197121  4446 Feb 15 01:34 LinalgBase.td
-rw-r--r-- 1 Martin 197121   974 Feb 15 01:34 LinalgDoc.td
-rw-r--r-- 1 Martin 197121 22334 Feb 15 02:38 LinalgOps.cpp.inc
-rw-r--r-- 1 Martin 197121  3022 Feb 15 01:34 LinalgOps.h
-rw-r--r-- 1 Martin 197121  9559 Feb 15 02:37 LinalgOps.h.inc
-rw-r--r-- 1 Martin 197121  9535 Feb 15 01:34 LinalgOps.td
-rw-r--r-- 1 Martin 197121 48142 Feb 15 02:37 LinalgStructuredOps.cpp.inc
-rw-r--r-- 1 Martin 197121 21014 Feb 15 02:37 LinalgStructuredOps.h.inc
-rw-r--r-- 1 Martin 197121 29586 Feb 15 01:34 LinalgStructuredOps.td
-rw-r--r-- 1 Martin 197121  3405 Feb 15 02:38 LinalgStructuredOpsInterfaces.cpp.inc
-rw-r--r-- 1 Martin 197121  9531 Feb 15 02:37 LinalgStructuredOpsInterfaces.h.inc
-rw-r--r-- 1 Martin 197121  9535 Feb 15 01:34 LinalgTraits.h
-rw-r--r-- 1 Martin 197121  1952 Feb 15 01:34 LinalgTypes.h

./Linalg/Transforms:
total 20
-rw-r--r-- 1 Martin 197121  679 Feb 15 02:38 LinalgTransformPatterns.h.inc
-rw-r--r-- 1 Martin 197121 5155 Feb 15 01:34 LinalgTransformPatterns.td
-rw-r--r-- 1 Martin 197121 4939 Feb 15 01:34 LinalgTransforms.h

./Linalg/Utils:
total 16
-rw-r--r-- 1 Martin 197121   997 Feb 15 01:34 Intrinsics.h
-rw-r--r-- 1 Martin 197121 11159 Feb 15 01:34 Utils.h

./LoopOps:
total 48
-rw-r--r-- 1 Martin 197121 19419 Feb 15 02:38 LoopOps.cpp.inc
-rw-r--r-- 1 Martin 197121  1663 Feb 15 01:34 LoopOps.h
-rw-r--r-- 1 Martin 197121  9122 Feb 15 02:38 LoopOps.h.inc
-rw-r--r-- 1 Martin 197121  9949 Feb 15 01:34 LoopOps.td

./QuantOps:
total 132
-rw-r--r-- 1 Martin 197121  3373 Feb 15 01:34 FakeQuantSupport.h
-rw-r--r-- 1 Martin 197121  1634 Feb 15 01:34 Passes.h
-rw-r--r-- 1 Martin 197121 50453 Feb 15 02:38 QuantOps.cpp.inc
-rw-r--r-- 1 Martin 197121  1278 Feb 15 01:34 QuantOps.h
-rw-r--r-- 1 Martin 197121 14767 Feb 15 02:38 QuantOps.h.inc
-rw-r--r-- 1 Martin 197121 10465 Feb 15 01:34 QuantOps.td
-rw-r--r-- 1 Martin 197121  2643 Feb 15 01:34 QuantPredicates.td
-rw-r--r-- 1 Martin 197121 17065 Feb 15 01:34 QuantTypes.h
-rw-r--r-- 1 Martin 197121  2745 Feb 15 01:34 QuantizeUtils.h
-rw-r--r-- 1 Martin 197121  9071 Feb 15 01:34 UniformSupport.h

./SDBM:
total 36
-rw-r--r-- 1 Martin 197121  7817 Feb 15 01:34 SDBM.h
-rw-r--r-- 1 Martin 197121  1010 Feb 15 01:34 SDBMDialect.h
-rw-r--r-- 1 Martin 197121 21847 Feb 15 01:34 SDBMExpr.h

./SPIRV:
total 2032
-rw-r--r-- 1 Martin 197121   2929 Feb 15 01:34 LayoutUtils.h
-rw-r--r-- 1 Martin 197121   1609 Feb 15 01:34 Passes.h
-rw-r--r-- 1 Martin 197121  16027 Feb 15 01:34 SPIRVArithmeticOps.td
-rw-r--r-- 1 Martin 197121  15303 Feb 15 01:34 SPIRVAtomicOps.td
-rw-r--r-- 1 Martin 197121   1152 Feb 15 02:38 SPIRVAvailability.cpp.inc
-rw-r--r-- 1 Martin 197121   6106 Feb 15 02:38 SPIRVAvailability.h.inc
-rw-r--r-- 1 Martin 197121   3374 Feb 15 01:34 SPIRVAvailability.td
-rw-r--r-- 1 Martin 197121 125464 Feb 15 01:34 SPIRVBase.td
-rw-r--r-- 1 Martin 197121   1799 Feb 15 01:34 SPIRVBinaryUtils.h
-rw-r--r-- 1 Martin 197121  18884 Feb 15 01:34 SPIRVBitOps.td
-rw-r--r-- 1 Martin 197121   9440 Feb 15 01:34 SPIRVCastOps.td
-rw-r--r-- 1 Martin 197121   5319 Feb 15 01:34 SPIRVCompositeOps.td
-rw-r--r-- 1 Martin 197121  14692 Feb 15 01:34 SPIRVControlFlowOps.td
-rw-r--r-- 1 Martin 197121   2718 Feb 15 01:34 SPIRVDialect.h
-rw-r--r-- 1 Martin 197121  83746 Feb 15 02:38 SPIRVEnumAvailability.cpp.inc
-rw-r--r-- 1 Martin 197121   5117 Feb 15 02:38 SPIRVEnumAvailability.h.inc
-rw-r--r-- 1 Martin 197121 122608 Feb 15 02:38 SPIRVEnums.cpp.inc
-rw-r--r-- 1 Martin 197121  45290 Feb 15 02:38 SPIRVEnums.h.inc
-rw-r--r-- 1 Martin 197121  17245 Feb 15 01:34 SPIRVGLSLOps.td
-rw-r--r-- 1 Martin 197121   2214 Feb 15 01:34 SPIRVGroupOps.td
-rw-r--r-- 1 Martin 197121  28949 Feb 15 01:34 SPIRVLogicalOps.td
-rw-r--r-- 1 Martin 197121   3608 Feb 15 01:34 SPIRVLowering.h
-rw-r--r-- 1 Martin 197121   2369 Feb 15 01:34 SPIRVNonUniformOps.td
-rw-r--r-- 1 Martin 197121 142965 Feb 15 02:38 SPIRVOpAvailabilityImpl.inc
-rw-r--r-- 1 Martin 197121   6977 Feb 15 02:38 SPIRVOpUtils.inc
-rw-r--r-- 1 Martin 197121 671676 Feb 15 02:38 SPIRVOps.cpp.inc
-rw-r--r-- 1 Martin 197121   1515 Feb 15 01:34 SPIRVOps.h
-rw-r--r-- 1 Martin 197121 243720 Feb 15 02:38 SPIRVOps.h.inc
-rw-r--r-- 1 Martin 197121  14494 Feb 15 01:34 SPIRVOps.td
-rw-r--r-- 1 Martin 197121 344899 Feb 15 02:38 SPIRVSerialization.inc
-rw-r--r-- 1 Martin 197121  15161 Feb 15 01:34 SPIRVStructureOps.td
-rw-r--r-- 1 Martin 197121   6688 Feb 15 01:34 SPIRVTypes.h
-rw-r--r-- 1 Martin 197121   1475 Feb 15 01:34 Serialization.h
-rw-r--r-- 1 Martin 197121  17666 Feb 15 02:38 TargetAndABI.cpp.inc
-rw-r--r-- 1 Martin 197121   2268 Feb 15 01:34 TargetAndABI.h
-rw-r--r-- 1 Martin 197121   1931 Feb 15 02:38 TargetAndABI.h.inc
-rw-r--r-- 1 Martin 197121   2649 Feb 15 01:34 TargetAndABI.td

./StandardOps:
total 432
-rw-r--r-- 1 Martin 197121 246994 Feb 15 02:38 Ops.cpp.inc
-rw-r--r-- 1 Martin 197121  13063 Feb 15 01:34 Ops.h
-rw-r--r-- 1 Martin 197121 103103 Feb 15 02:38 Ops.h.inc
-rw-r--r-- 1 Martin 197121  60748 Feb 15 01:34 Ops.td
-rw-r--r-- 1 Martin 197121   2180 Feb 15 02:38 OpsEnums.cpp.inc
-rw-r--r-- 1 Martin 197121   1748 Feb 15 02:38 OpsEnums.h.inc

./Utils:
total 4
-rw-r--r-- 1 Martin 197121 4068 Feb 15 01:34 StructuredOpsUtils.h

./VectorOps:
total 212
-rw-r--r-- 1 Martin 197121   5224 Feb 15 01:34 Utils.h
-rw-r--r-- 1 Martin 197121 101308 Feb 15 02:38 VectorOps.cpp.inc
-rw-r--r-- 1 Martin 197121   2205 Feb 15 01:34 VectorOps.h
-rw-r--r-- 1 Martin 197121  39809 Feb 15 02:38 VectorOps.h.inc
-rw-r--r-- 1 Martin 197121  45139 Feb 15 01:34 VectorOps.td
-rw-r--r-- 1 Martin 197121    679 Feb 15 02:38 VectorTransformPatterns.h.inc
-rw-r--r-- 1 Martin 197121   1006 Feb 15 01:34 VectorTransformPatterns.td
-rw-r--r-- 1 Martin 197121   3292 Feb 15 01:34 VectorTransforms.h
```

Comment (Robertorosmaninho) [reply cmake/Modules/FindMLIR.cmake]: Those are the files generated by MLIR itself, this function is called to generate the files from D Dialect. The Ops.td, the file with the description of D Dialect operation,  has not yet been merged.

**Comment by kinke** [comment]
> The Ops.td, the file with the description of D Dialect operation, has not yet been merged.

Can you link to some PR/GitHub branch? If MLIR itself needs to be modified to include some 'D dialect', it seems way easier to integrate that into our LLVM fork than hacking CMake here and depending on a MLIR source tree, which won't work for most people.

File: cmake/Modules/FindMLIR.cmake (hunk 4 above)
Comment (Robertorosmaninho) [reply cmake/Modules/FindMLIR.cmake]: I solved this and [this](https://github.com/ldc-developers/ldc/pull/3312#issuecomment-586596330) problems, can I send another PR to fix it?

Comment (kinke) [reply cmake/Modules/FindMLIR.cmake]: Sure, please go ahead.

**Comment by kinke** [comment]
Just as a heads up, if this isn't cleaned up by v1.21 final, I'll revert this.

**Comment by Robertorosmaninho** [comment]
> > The Ops.td, the file with the description of D Dialect operation, has not yet been merged.
> 
> Can you link to some PR/GitHub branch? If MLIR itself needs to be modified/extended to include some 'D dialect', it seems way easier to integrate that into our LLVM fork than hacking CMake here and depending on a MLIR source tree, which won't work for most people.

The modifications that I propose are not in MLIR itself, so I don't think that it's possible to edit our LLVM fork to work that way. The Dialects of MLIR are just the basic language of the framework, the idea is each frontend develop it's own Dialect, not as an extension of MLIR, but of the frontend to interact with MLIR. It's important to say that MLIR is a dependence only if the LLVM project is built this MLIR enabled if don't, ldc works as should work without it.

**Comment by kinke** [comment]
What are the exact files that you need from the MLIR source tree and aren't available in a regular LLVM+MLIR installation (https://github.com/ldc-developers/llvm-project/releases/tag/CI)?

**Comment by Robertorosmaninho** [comment]
That's is the LLVM 8 distribution

**Comment by kinke** [comment]
This is pretty recent LDC-LLVM v10.0.0-rc4...

**Comment by JohanEngelen** [comment]
There should not be any need for a custom LLVM/MLIR build, nor modifying any files in MLIR source tree.
The "D dialect" should be described by a file in LDC's sourcetree, I propose to put it in `./mlir/` directory (new).

**Comment by Robertorosmaninho** [comment]
> What are the exact files that you need from the MLIR source tree and aren't available in a regular LLVM+MLIR installation (https://github.com/ldc-developers/llvm-project/releases/tag/CI)?

Do you mean header files? They are available in this distribution. I don't modify anything in MLIR tree, just need the files as includes.

**Comment by Robertorosmaninho** [comment]
> There should not be any need for a custom LLVM/MLIR build, nor modifying any files in MLIR source tree.

That's right, these modifications are just necessary to enable the MLIR headers as includes in LDC files.

> The "D dialect" should be described by a file in LDC's sourcetree, I propose to put it in `./mlir/` directory (new).

Yes, I created this mlir directory and put the files I created to convert AST to MLIR in that ./gen/mlir directory, as shown [here](https://github.com/Robertorosmaninho/ldc/tree/Introducing-MLIR-translation)

**Comment by kinke** [comment]
> Do you mean header files? They are available in this distribution. I don't modify anything in MLIR tree,I just need the files as includes.

Never mind, glad Johan engaged to clean up this mess.

**Comment by Robertorosmaninho** [comment]
> > Do you mean header files? They are available in this distribution. I don't modify anything in MLIR tree,I just need the files as includes.
> 
> Never mind, glad Johan engaged to clean up this mess.

thanks.

**Pull request merged**